drivers/sensor/bmp388/bmp388.c

```diff
@@ -0,0 +1,691 @@
+	bmp388->cs_ctrl.delay = 0;
+#endif /* DT_INST_SPI_DEV_HAS_CS_GPIOS(0) */
+
+	/* reboot the chip */
+	if (bmp388_byte_write(dev, BMP388_REG_CMD, BMP388_CMD_SOFT_RESET) < 0) {
+		LOG_ERR("Cannot reboot chip.");
+		return -EIO;
+	}
+
+	k_busy_wait(2000);
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+	/* do a dummy read from 0x7F to activate SPI */
+	if (bmp388_byte_read(dev, 0x7F, &val) < 0) {
+		return -EIO;
+	}
+
+	k_busy_wait(100);
+#endif
+
+	if (bmp388_byte_read(dev, BMP388_REG_CHIPID, &val) < 0) {
+		LOG_ERR("Failed to read chip id.");
+		return -EIO;
+	}
+
+	if (val != BMP388_CHIP_ID) {
+		LOG_ERR("Unsupported chip detected (0x%x)!", val);
+		return -ENODEV;
+	}
+
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	bmp388->device_power_state = DEVICE_PM_ACTIVE_STATE;
+#endif
+
+	/* Read calibration data */
+	if (bmp388_get_calibration_data(dev) < 0) {
+		LOG_ERR("Failed to read calibration data.");
+		return -EIO;
+	}
+
+	/* Set ODR */
+	if (bmp388_reg_field_update(dev,
+				    BMP388_REG_ODR,
+				    BMP388_ODR_MASK,
+				    bmp388->odr) < 0) {
+		LOG_ERR("Failed to set ODR.");
+		return -EIO;
+	}
+
+	/* Set OSR */
+	val = (bmp388->osr_pressure << BMP388_OSR_PRESSURE_POS);
+	val |= (bmp388->osr_temp << BMP388_OSR_TEMP_POS);
+	if (bmp388_byte_write(dev, BMP388_REG_OSR, val) < 0) {
+		LOG_ERR("Failed to set OSR.");
+		return -EIO;
+	}
+
+	/* Enable sensors and normal mode*/
+	if (bmp388_byte_write(dev,
+			      BMP388_REG_PWR_CTRL,
+			      BMP388_PWR_CTRL_ON) < 0) {
+		LOG_ERR("Failed to enable sensors.");
+		return -EIO;
+	}
+
+#ifdef CONFIG_BMP388_TRIGGER
+	if (bmp388_trigger_mode_init(dev) < 0) {
+		LOG_ERR("Cannot set up trigger mode.");
+		return -EINVAL;
+	}
+#endif
+
+	return 0;
+}
+
+struct bmp388_data bmp388_data = {
+	.odr = BMP388_DEFAULT_ODR,
+	.osr_pressure = BMP388_DEFAULT_OSR_PRESSURE,
+	.osr_temp = BMP388_DEFAULT_OSR_TEMP,
+};
+
+const struct bmp388_config bmp388_config = {
+	.bus_name = DT_INST_BUS_LABEL(0),
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+	.bus_addr = DT_INST_REG_ADDR(0),
+#endif
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+	.spi_cfg.frequency = DT_INST_PROP(0, spi_max_frequency),
+	.spi_cfg.operation = SPI_OP_MODE_MASTER | SPI_WORD_SET(8),
+	.spi_cfg.slave = DT_INST_REG_ADDR(0),
+#if DT_INST_SPI_DEV_HAS_CS_GPIOS(0)
+	.cs_port = DT_INST_SPI_DEV_CS_GPIOS_LABEL(0),
+	.cs_pin = DT_INST_SPI_DEV_CS_GPIOS_PIN(0),
+	.cs_flags = DT_INST_SPI_DEV_CS_GPIOS_FLAGS(0),
+	.spi_cfg.cs = &bmp388_data.cs_ctrl,
+#else
+	.spi_cfg.cs = NULL,
+#endif  /* DT_INST_SPI_DEV_HAS_CS_GPIOS(0) */
+#endif  /* DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) */
+
+#ifdef CONFIG_BMP388_TRIGGER
+	.int_port = DT_INST_GPIO_LABEL(0, int_gpios),
+	.int_pin = DT_INST_GPIO_PIN(0, int_gpios),
+	.int_flags = DT_INST_GPIO_FLAGS(0, int_gpios),
+#endif
+};
+
+DEVICE_DEFINE(
+	bmp388,
+	DT_INST_LABEL(0),
+	bmp388_init,
+	bmp388_device_ctrl,
+	&bmp388_data,
+	&bmp388_config,
+	POST_KERNEL,
+	CONFIG_SENSOR_INIT_PRIORITY,
+	&bmp388_api);
```

Comment:
I think this driver should be made multi-instance.

Reply:
I took a pass at this a while back but the ```#if``` blocks in the definition of the ```bmp388_config``` struct make it hard to wrap this all in a macro. Many sensors drivers are currently single instance, is this a hard requirement?

Reply:
> Many sensors drivers are currently single instance, is this a hard requirement?

That's up to @MaureenHelm as the sensor maintainer.

Reply:
We are slightly moving all sensors, or at least the most recent one, to be multi-instance. I would say that this should be the rule for every new sensor implementation. I leave to @mhelm the last word on it.

Reply:
Got it, I've push a change that makes this multi-instance aware.